packages/teams-js/src/private/externalAppAuthenticationForCEA.ts

```diff
@@ -0,0 +1,247 @@
+   */
+  export async function authenticateWithOauth(
+    appId: AppId,
+    conversationId: string,
+    authenticateParameters: externalAppAuthentication.AuthenticatePopUpParameters,
+  ): Promise<void> {
+    ensureInitialized(runtime, FrameContexts.content);
+
+    if (!isSupported()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    validateId(conversationId, new Error('conversation id is not valid.'));
+
+    // Ask the parent window to open an authentication window with the parameters provided by the caller.
+    const error = await sendAndUnwrap<externalAppAuthentication.InvokeError | undefined>(
+      getApiVersionTag(
+        externalAppAuthenticationTelemetryVersionNumber,
+        ApiName.ExternalAppAuthenticationForCEA_AuthenticateWithOauth,
+      ),
+      ApiName.ExternalAppAuthenticationForCEA_AuthenticateWithOauth,
+      appId.toString(),
+      conversationId,
+      authenticateParameters.url.href,
+      authenticateParameters.width,
+      authenticateParameters.height,
+      authenticateParameters.isExternal,
+    );
+    if (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * @beta
+   * @hidden
+   * Signals to the host to perform authentication using the given authentication parameters and then resend the request to the application backend with the authentication result.
+   * @internal
+   * Limited to Microsoft-internal use
+   * @param appId App ID of the app upon whose behalf Copilot is requesting authentication. This must be a UUID.
+   * @param conversationId ConversationId To tell the bot what conversation the calls are coming from
+   * @param authenticateParameters Parameters for the authentication pop-up
+   * @param originalRequestInfo Information about the original request that should be resent
+   * @throws InvokeError if the host encounters an error while authenticating or resending the request
+   * @returns A promise that resolves to the IActionExecuteResponse from the application backend and rejects with InvokeError if the host encounters an error while authenticating or resending the request
+   */
+  export async function authenticateAndResendRequest(
+    appId: AppId,
+    conversationId: string,
+    authenticateParameters: externalAppAuthentication.AuthenticatePopUpParameters,
+    originalRequestInfo: externalAppAuthentication.IActionExecuteInvokeRequest,
+  ): Promise<externalAppAuthentication.IActionExecuteResponse> {
+    ensureInitialized(runtime, FrameContexts.content);
+
+    if (!isSupported()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    validateId(conversationId, new Error('conversation id is not valid.'));
+
+    validateOriginalRequestInfo(originalRequestInfo);
+
+    // Ask the parent window to open an authentication window with the parameters provided by the caller.
+    const [response] = await sendMessageToParentAsync<
```

Comment:
>sendMessageToParentAsync [](http://example.com/codeflow?start=29&length=24)

Can this also use `sendAndUnwrap` to avoid needing the seemingly unnecessary array?

Reply:
(below this, too)

Reply:
Fixed!